metasynthesis/abstract_genetic.py

```diff
@@ -0,0 +1,67 @@
+from typing import List, Callable, Tuple
+
+Genome = List[int]
+Population = List[Genome]
+PopulateFunc = Callable[[], Population]
+FitnessFunc = Callable[[Genome], float]
+SelectionFunc = Callable[[Population, FitnessFunc], Tuple[Genome, Genome]]
+CrossoverFunc = Callable[[Genome, Genome], Tuple[Genome, Genome]]
+MutationFunc = Callable[[Genome], Genome]
+
+
+class GeneticAlgorithm:
+    """Abstract interface for a genetic algorithm to synthesise elements used in program synthesis"""
+
+    def __init__(self, fitness_limit: int, generation_limit: int):
+        self.fitness_limit = fitness_limit
+        self.generation_limit = generation_limit
+
+    def generate_genome(self, length: int) -> Genome:
+        """This method creates a new genome of the specified length"""
+
+        raise NotImplementedError()
+
+    def generate_population(self, size: int, genome_length: int) -> Population:
+        """This method creates a population of new genomes of the specified size"""
+
+        return [self.generate_genome(genome_length) for _ in range(size)]
+
+    def fitness(self, genome: Genome) -> float:
+        """This method calculates the fitness of the specified genome"""
+
+        raise NotImplementedError()
+
+    def single_point_crossover(self, a: Genome, b: Genome) -> Tuple[Genome, Genome]:
+        """This method combines first half of genome a with second half of b and vice versa"""
+
+        raise NotImplementedError()
+
+    def n_point_crossover(self, a: Genome, b: Genome) -> Tuple[Genome, Genome]:
+        """This method applies multiple point crossover"""
+
+        raise NotImplementedError()
+
+    def mutation(self, genome: Genome, probability: float) -> Genome:
```

Comment:
Maybe a good idea would be adding some function here as a parameter representing the kind of mutation?

Reply:
I guess for people that are planning on implementing a lot of big mutation methods that's indeed clearer